### ISSUE TITLE
[REVIEW]: batchtools: Tools for R to work on batch systems

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#135]
**Submitting author:** @mllg (<a href="http://orcid.org/0000-0001-9754-0393">Michel Lang</a>)
**Repository:** <a href="https://github.com/mllg/batchtools" target ="_blank">https://github.com/mllg/batchtools</a>
**Version:** v0.9.2
**Editor:** @arfon
**Reviewer:** @masalmon
**Archive:** <a href="http://dx.doi.org/10.5281/zenodo.293835" target="_blank">10.5281/zenodo.293835</a>

## Status

[![status](http://joss.theoj.org/papers/8eb56c3dcaedd5975ab6370eae446a8d/status.svg)](http://joss.theoj.org/papers/8eb56c3dcaedd5975ab6370eae446a8d)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/8eb56c3dcaedd5975ab6370eae446a8d"><img src="http://joss.theoj.org/papers/8eb56c3dcaedd5975ab6370eae446a8d/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/8eb56c3dcaedd5975ab6370eae446a8d/status.svg)](http://joss.theoj.org/papers/8eb56c3dcaedd5975ab6370eae446a8d)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/mllg/batchtools" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

### Conflict of interest

- [x] As the reviewer I confirm that there are no conflicts of interest for me to review this work (such as being a major contributor to the software).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/mllg/batchtools">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v0.9.0)?
- [x] **Authorship:** Has the submitting author (@mllg) made major contributions to the software?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** Have any performance claims of the software been confirmed?

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g. API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

Paper PDF: [10.21105.joss.00135.pdf](https://github.com/openjournals/joss-reviews/files/639511/10.21105.joss.00135.pdf)

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g. papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon. I'm here to help you with some common editorial tasks for JOSS.

For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by arfon** [comment]
👋  @masalmon - please carry out your review in this issue by updating the checklist above and giving feedback in this issue.

Any questions/concerns please let me know.

**Comment by masalmon** [comment]
Review

This is a really nice package! I think that I cannot realize how good it is because I've no parallel computing experience, but I can see it is a very complete package. @arfon by the way would it make sense to have a second review by someone a bit more experienced in the applications of such a package? Maybe the authors of the issues https://github.com/mllg/batchtools/issues/30 and https://github.com/mllg/batchtools/issues/44 ?

The software documentation / paper only needs minor revisions in my opinion. I have opened quite a few issues in the repo itself. Most of them are _suggestions_ regarding the organization of the documentation. My only real concern is the changes between the two packages as described in the JSS paper and the single package described in this JOSS submission: I think this deserves a bit more than pointing readers to the NEWS.md.

### Conflict of interest

- [ x] As the reviewer I confirm that there are no conflicts of interest for me to review this work (such as being a major contributor to the software).
### General checks

- [ x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/mllg/batchtools">repository url</a>?
- [ x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ x] **Version:** Does the release version given match the GitHub release (v0.9.0)?

Yes, but I wonder if v0.9.1 will the the version used for the JOSS paper in the end?
- [ x] **Authorship:** Has the submitting author (@mllg) made major contributions to the software?

### Functionality

- [x ] **Installation:** Does installation proceed as outlined in the documentation?
- [ x] **Functionality:** Have the functional claims of the software been confirmed?
- [ x] **Performance:** Have any performance claims of the software been confirmed?

### Documentation

- [x ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?

Just a small remark here: I guess you assume your users to know which e.g. scheduler to choose? Or would it make sense to point them to resources for choosing one?

- [x ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g. API method documentation)?

The reason I do not check this is because of the JSS paper remarks I mention at the beginning of this review.

- [x ] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

Except for the line " It would be great if you would help expand this collection to cover more exotic configurations. To do so, please send your template via mail or open a new pull request." in the Setup vignette, I did not see any contributions guideline.

### Software paper

Paper PDF: [10.21105.joss.00135.pdf](https://github.com/openjournals/joss-reviews/files/639511/10.21105.joss.00135.pdf)

- [ x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [ x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ x] **References:** Do all archival references that should have a DOI list one (e.g. papers, datasets, software)?

**Comment by arfon** [comment]
> This is a really nice package! I think that I cannot realize how good it is because I've no parallel computing experience, but I can see it is a very complete package. @arfon by the way would it make sense to have a second review by someone a bit more experienced in the applications of such a package? Maybe the authors of the issues mllg/batchtools#30 and mllg/batchtools#44 ?

Thanks for this review @masalmon. I agree it might make sense to solicit a second review here.

@HenrikBengtsson @jakob-r - as users of this package would either of you be willing to review this submission for JOSS? The process is very lightweight (basically going through the checklist above).

**Comment by mllg** [comment]
@jakob-r has a conflict of interest. He is a coworker of mine.

**Comment by arfon** [comment]
> @jakob-r has a conflict of interest. He is a coworker of mine.

Understood. Thanks for the heads up.

**Comment by HenrikBengtsson** [comment]
I'd be happy to review, but (regardless how easy it is) I will _not_ be able to get to this until January.

PS. I haven't started to use batchtools yet, but am definitely planning to dive into the details, cf. https://github.com/HenrikBengtsson/future.batchtools

**Comment by arfon** [comment]
> I'd be happy to review, but (regardless how easy it is) I will not be able to get to this until January.

👍 thanks for the heads up @HenrikBengtsson - that should be fine.

**Comment by masalmon** [comment]
I think most my comments about documentation have been taken into account now, great job. Only the contribution guidelines are missing or maybe I've missed something :slightly_smiling_face: 

The only other remaining issue I opened is [this one](https://github.com/mllg/batchtools/issues/55) which was only a suggestion.

**Comment by mllg** [comment]
I've fixed the last issue/suggestion and added contribution guidelines to the README. Waiting for @HenrikBengtsson now.

Thanks again @masalmon for your comments which substantially improved the package.

**Comment by masalmon** [comment]
Awesome @mllg, it all looks good to me. I'm looking forward to reading the next review & the corresponding commits. :grinning:

**Comment by arfon** [comment]
Friendly reminder @HenrikBengtsson 😁

**Comment by HenrikBengtsson** [comment]
Haven't forgot.

**Comment by HenrikBengtsson** [comment]
I've now completed my code / document review resulting in the following batchtools issues:

* https://github.com/mllg/batchtools/issues/69
* https://github.com/mllg/batchtools/issues/70
* https://github.com/mllg/batchtools/issues/73
* https://github.com/mllg/batchtools/issues/74
* https://github.com/mllg/batchtools/issues/75
* https://github.com/mllg/batchtools/issues/76
* https://github.com/mllg/batchtools/issues/77
* https://github.com/mllg/batchtools/issues/78
* https://github.com/mllg/batchtools/issues/79
* https://github.com/mllg/batchtools/issues/80
* https://github.com/mllg/batchtools/issues/81

Most of these are minor and not critical for accepting the manuscript.  The only one that could be consider **major** work / thought is [Issue 81](https://github.com/mllg/batchtools/issues/81) on 'RANDOMNESS: L'Cuyer RNG streams instead of ad-hoc random seeds?', which I think is an important topic.  I'd love to hear feedback from others who know / appreciate the problems of random number generation (and reproducibility) in parallel systems.

I've tested batchtools successfully on a local Linux machine, using the basic backends ("cluster functions") like "interactive" and "socket", which should work out of the box for everyone regardless of OS.   Now, I'm in the process of testing batchtools on a TORQUE / PBS system, on which I've previously used authors' BatchJobs package successfully / extensively.  I'm currently running into a few hiccups with batchtools ("Warning in waitForJobs() : Some jobs disappeared from the system") that I need to troubleshoot locally before reaching out / drawing any conclusions.  I would like to see a successful run on a real HPC cluster myself, although I'm in no doubt that the authors would post-publication make it work if it turns out that there is a bug.

Q to journal: "Authors: Does the paper.md file include a list of authors with their affiliations?" - I only see a link to the PDF, where's this Markdown file?  Also, in general, is there meant to be a separate issue / repos for discussing parts of a submitted paper, or is that meant to be done over at authors' software repos?

**Comment by arfon** [comment]
@HenrikBengtsson - many thanks for the review. You've not updated the checklist at the top of this issue - is that because you don't have (GitHub) permissions to do this?

> Q to journal: "Authors: Does the paper.md file include a list of authors with their affiliations?" - I only see a link to the PDF, where's this Markdown file? Also, in general, is there meant to be a separate issue / repos for discussing parts of a submitted paper, or is that meant to be done over at authors' software repos?

The `paper.md` file is here: https://github.com/mllg/batchtools/blob/master/paper/paper.md . If there are issues with this then you should post them on the repository (https://github.com/mllg/batchtools) too.

**Comment by arfon** [comment]
Also, @mllg - let me know when you've managed to respond to @HenrikBengtsson's feedback.

**Comment by HenrikBengtsson** [comment]
@arfon, @masalmon cut'n'paste the checklist in a comment - shouldn't I do the same?  Who decides to check the top one? (given more than one reviewer?)

**Comment by arfon** [comment]
> @arfon, @masalmon cut'n'paste the checklist in a comment - shouldn't I do the same? Who decides to check the top one? (given more than one reviewer?)

Good point. I forgot there had been a review already 😄 

I think it's fine for you to cut and paste in this case like @masalmon did already.

**Comment by HenrikBengtsson** [comment]
UPDATES:

Additional feedback based on some serious testing on real compute clusters:

* https://github.com/mllg/batchtools/issues/85
* https://github.com/mllg/batchtools/issues/87
* https://github.com/mllg/batchtools/issues/88

which @mmlg is addressing.

I've now successfully tested batchtools 0.9.1-9000 with:

* `makeClusterFunctionsInteractive(external = TRUE)` on multiple system including the below two clusters
* `makeClusterFunctionsTORQUE()` on a large TORQUE / PBS cluster running RedHat 6.6 and R 3.3.2
* `makeClusterFunctionsSGE()`: on a large Son of Grid Engine (SGE) cluster running Centos 7.3 and R 3.3.2

using some rather extensive real-world computations.

I'm extremely pleased with the experience I now have with batchtools.  **It's a very nice piece of work and extremely valuable!**  Thank you.

As soon as the developer's version batchtools 0.9.1-9000 is wrapped up and the updated version is submitted to CRAN, I'll be happy to recommend anyone to start using it for their HPC needs.  I also feel comfortable enough about this upcoming version of batchtools, that'll be recommending existing BatchJobs users to migrate over to batchtools (as the authors have suggested for a while now).

I hereby strongly recommend that the JOSS paper should be accepted.

**Comment by HenrikBengtsson** [comment]
Also, @arfon, all the checkboxes at the top can be ticked. I don't have the permission to do it.

**Comment by mllg** [comment]
Thanks again @HenrikBengtsson for your thorough review. I will address the remaining small issues before Monday. 

@arfon What is the supposed procedure to change the version number in this review (0.9.0) to match the version of the next package release (0.9.2)?

**Comment by HenrikBengtsson** [comment]
Just to clarify, none of the remaining batchtools issues should prevent the acceptance of the article; they can be dealt with later and are non-critical.

**Comment by arfon** [comment]
> Just to clarify, none of the remaining batchtools issues should prevent the acceptance of the article; they can be dealt with later and are non-critical.

Thanks for the clarification.

> @arfon What is the supposed procedure to change the version number in this review (0.9.0) to match the version of the next package release (0.9.2)?

Just to let me know - I've made the change at the top of the thread here.

@mllg - Once you're ready, could you make an archive of the reviewed software in Zenodo/figshare/other service and update this thread with the DOI of the archive? I can then move forward with accepting the submission.

**Comment by mllg** [comment]
I've published a new release on [CRAN](https://cran.r-project.org/package=batchtools) and [Github/Zenodo](https://zenodo.org/record/293835).

DOI is <http://doi.org/10.5281/zenodo.293835>.

**Comment by arfon** [comment]
@whedon set 10.5281/zenodo.293835 as archive

**Comment by whedon** [comment]
OK. <a href="http://dx.doi.org/10.5281/zenodo.293835" target="_blank">10.5281/zenodo.293835</a> is the archive.

**Comment by arfon** [comment]
Many thanks for your reviews @masalmon and @HenrikBengtsson ✨ 

@mllg - your paper is now accepted into JOSS and your DOI is http://dx.doi.org/10.21105/joss.00135 :zap: 🚀 💥
